### DISCUSSION TITLE
Ray-Optics v0.7.0

### DISCUSSION THREAD

**Discussion by mjhoptics** [mjhoptics/ray-optics#47] [General]
A new version of `ray-optics` has been uploaded. There have been numerous internal, structural, changes made to the code. Additionally, the required versions for a number of packages (e.g. numpy, scipy, matplotlib, etc.) have been updated. I have tried to ensure scripts and models are backwards compatible, but it's software so undoubtedly there will be issues. I recommend installing it in a new virtual environment just to be on the safe side.

Improvements:
---------------

- Add fractional field height specification to `raytr.opticalspec.FieldSpec`.
- Add semi-diameter to list of items in `seq.sequential.SequentialModel.add_surface()` method.
- Add `elem.elements.CementedElement` element to `elem.elements` module.
- Add `seq.sequential.SequentialModel.reverse_path()` method to facilitate reverse ray tracing through the model.
- Add a mapping interface to `optical.opticalmodel.OpticalModel` and `raytr.opticalspec.OpticalSpecs`; allows submodel access like opm['sm'] and opm['osp]['fov'].
- Add `elem.parttree.PartTree` to `optical.opticalmodel.OpticalModel`. Many changes to get `SequentialModel`, `ElementModel` and `PartTree` to work in sync. See below.

Bug Fixes:
-----------

- Fix paraxial image distance calculation.
- Fix DOE implementation for DOE on exiting side of element.
- Add exception handling for ray trace errors in `raytr.analyses` module.

Many miscellaneous fixes and improvements.

Part Tree:
----------
Describing optical systems as a sequence of optical surfaces is an extremely compact and efficient way to model a system for ray tracing. But this is a simplification for a specific purpose and other needs, such as editing the system prescription, are easily expressed with an element model. The y-ybar diagram is another example where sometimes it is useful for each node to map to a surface and other times where having each node map to an element or even a subsystem is more advantageous. 

Ray-optics has had an element model from the beginning; the need to draw lens layouts makes the element description essential. The goal of the `InteractiveLayout` view, when used in the Qt frontend, is to allow editing in the element space, e.g. sliding elements, tipping mirrors, etc, and have those actions correctly map to the sequential model. The goal is to be able to change the model via either the sequential, element, or paraxial models.

The Part Tree is a tree structure that connects the element model objects to the sequential model interfaces and gaps. Every node in the tree has a tag. The `PartTree` method nodes_with_tag() will return a list of nodes corresponding to the tag(s). The query

`opm['part_tree'].node_with_tag(tag='#element')`

will return a list of all of the elements (mirrors, lenses and cemented lenses) in the model.

`opm['part_tree'].node_with_tag(tag=''#element#airgap#dummyifc'')`

will return elements plus airgaps and dummy interfaces.

To determine which element a particular interface is related to, use the parent_node() query:

`opm['part_tree'].parent_object(sm.ifcs[5])`

returns the element (and it's node) that surface 5 in the sequential model is associated with.

Building this tree structure for a model allows the validity of the surface model to be checked. This is of particular importance for modeling double pass elements or systems. The sequential model in isolation doesn't have the context that some surfaces may be appearing multiple times in the model. Having an accurate element model is the first step to automated validation of model changes.

Mapping interface for OpticalModel and Optical_Spec:
-------------------------------------------------------

`OpticalModel` and `Optical_Spec` instances are containers for other models. To date, the models contained have been accessible directly, as they are public attributes. A mapping or `dict` style interface has been added. There is some benefit to this approach when saving and restoring ray-optics files from different versions. It also recognizes that additional submodels will probably be added in the future, possibly by other users. Syntactically, I like it.

The keys for the `OpticalModel` are:

```
'seq_model', 'sm'
'optical_spec', 'osp'
'ele_model', 'em'
'parax_model', 'pm'
'part_tree', 'pt'
'system_spec', 'sys'
'specsheet', 'ss'
```

The keys for `Optical_Spec` are:

```
'wvls'
'pupil'
'fov'
'focus'
```

Deprecating the direct attribute access is being considered.

Near Term Plans
----------------

I am presenting a paper on `ray-optics` at the [International Optical Design Conference](https://www.osa.org/en-us/meetings/osa_meetings/optical_design_and_fabrication/program/international_optical_design/) at the end of June. This will keep me distracted until then. I will attempt to fix issues relating to the new version as quickly as possible; please submit issues as needed. 

The ability to trace ray fans, lists and grids in the `raytr.analyses` module needs to be integrated with the lens layout capability. This will help greatly in understanding what is going on in a system.

The ability to switch between surface and element modes in the y-ybar diagram is in work. It has real promise but also has raises some issues that require research.

The element model can be used, with some additional data, to do a rudimentary form of non-sequential ray tracing in order to determine a sequence of interfaces through a set of elements. This would allow automatically generating sequential models for double pass systems, for instance.

Contributors
-------------

I'd like to acknowledge several users who have taken the time to document issues with ray-optics and helping me make it a better piece of software. @dibyendumajumdar has created a number of notebooks in the ray-optics-notebooks repo, based on prescriptions from William Claff's Optical Bench app at www.photonstophotos.net. Thank you for your interest and your work.

Feel free to comment on this. Feel free to use the Discussion forum for other questions and use Issues for problem reports.

Thank you!

Mike

**Comment by dibyendumajumdar** [comment]
Hi Mike,

Thank you for the fantastic work you are doing. Its inspiring other projects I can see.

> I am presenting a paper on ray-optics at the International Optical Design Conference at the end of June. This will keep me distracted until then. I will attempt to fix issues relating to the new version as quickly as possible; please submit issues as needed.

Would be nice if you shared this here - will it be recorded?

Regards
Dibyendu

**Reply by mjhoptics** [reply]
Depending on copyright restrictions, I'll either post it or sneak you a link ;-)

**Reply by sonebu** [reply]
If the link-sneaking isn't over, I'd love to join :D
